ang/civiawards-workflow/services/applicant-management-workflow.service.js

```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * @param {object} scope scope object of thw workflow list controller
+     * @returns {Promise} Promise which resolved to a list of case type ids
+     */
+    function applyFilter (scope) {
+      return processMyAwardsFilter(scope.selectedFilters.awardFilter)
+        .then(function (caseTypeIDs) {
+          var filters = prepareFilterObject(scope, caseTypeIDs);
+
+          return civicaseCrmApi('AwardDetail', 'get', filters)
+            .then(function (awardsData) {
+              return _.map(awardsData.values, function (award) {
+                return award.case_type_id;
+              });
+            });
+        });
+    }
+
+    /**
+     * Returns workflows list for applicant management
+     *
+     * @param {Array} scope scope object of the workflows list controller
+     * @returns {Array} api call parameters
+     */
+    function getWorkflowsList (scope) {
+      return applyFilter(scope)
+        .then(function (caseTypeIds) {
+          var params = {
+            sequential: 1,
+            is_active: scope.selectedFilters.is_active,
+            case_type_category: scope.caseTypeCategory,
+            options: { limit: 0 },
+            'api.AwardDetail.get': {
+              case_type_id: '$value.id'
+            }
+          };
+
+          if (caseTypeIds) {
+            if (caseTypeIds.length === 0) {
+              return $q.resolve([]);
+            }
+            params.id = { IN: caseTypeIds };
+          }
+
+          return civicaseCrmApi('CaseType', 'get', params).then(function (data) {
+            return getFormattedAwardDetailsData(data.values);
+          });
+        });
+    }
+
+    /**
+     * @param {object} workflow workflow object
+     * @returns {Promise} promise
+     */
+    function createDuplicate (workflow) {
+      return civicaseCrmApi('CaseType', 'create', _.extend({}, workflow, { id: null }))
+        .then(function (data) {
+          var awardsDetailsData = _.clone(workflow.awardDetails);
+
+          awardsDetailsData.id = null;
+          awardsDetailsData.case_type_id = data.values[0].id;
+
+          return civicaseCrmApi([
+            ['AwardDetail', 'create', awardsDetailsData]
+          ]);
+        });
+    }
+
+    /**
+     * @param {object[]} workflows list of workflows
+     * @returns {object[]} formatted list of workflows
+     */
+    function getFormattedAwardDetailsData (workflows) {
+      var managerContacts = [];
+
+      workflows = _.map(workflows, function (workflow) {
+        workflow.awardDetails = workflow['api.AwardDetail.get'].values[0];
+        delete workflow['api.AwardDetail.get'];
+
+        managerContacts = managerContacts.concat(workflow.awardDetails.award_manager);
+
+        workflow.awardDetailsFormatted = {
+          subtypeLabel: awardSubtypes[workflow.awardDetails.award_subtype].label,
+          managers: []
+        };
+
+        return workflow;
+      });
+
+      return ContactsCache.add(managerContacts)
```

Comment:
By calling `ContactsCache.add(managerContacts)` from inside a `.map` we are going to execute this function multiple times. We should try to group all the managers for all the workflows first, and then call this function.

Reply:
This was not called inside a map. It was only called once.